File: pi-steer/pi_steer/agio.py
import time
import pi_steer.debug as db
import pi_steer.section_control
import socket
import struct
import threading

_HELLO = 0xc7
_LATLON = 0xd0
_AGIOTRAFFIC = 0xd2
_IMU = 0xd3
_IMU_DETACH_REQ = 0xd4
_NMEA_BYTES = 0xd5
_SWITCH_CONTROL = 0xea
_MACHINE_CONFIG = 0xee
_RELAY_CONFIG = 0xec
_MACHINE_DATA = 0xef
_STEER_CONFIG = 0xfb
_STEERSETTINGS = 0xfc
_FROM_AUTOSTEER = 0xfd
_AUTOSTEER_DATA = 0xfe


source_text = {
    0x7b: 'switch control',
    0x7c: 'GPS',
    0x7d: 'IMU',
    0x7e: 'Autosteer',
    0x7f: 'AgIO',
}

pgn_text = {
    _HELLO: 'Hello???',
    _LATLON: 'LatLon',
    _AGIOTRAFFIC: 'AgIOTraffic',
    _IMU: 'IMU',
    _IMU_DETACH_REQ: 'IMU Detach Req',
    _NMEA_BYTES: 'NMEA bytes',
    _SWITCH_CONTROL: 'switchControl',
    _MACHINE_CONFIG: 'machineConfig',
    _RELAY_CONFIG: 'relayConfig',
    _MACHINE_DATA: 'machineData',
    _STEER_CONFIG: 'steerConfig',
    _STEERSETTINGS: 'steerSettings',
    _FROM_AUTOSTEER: 'fromAutoSteer',
    _AUTOSTEER_DATA: 'autoSteerData',
}

pgn_data = {
    _HELLO: # Hello???
    lambda data: { 
    },
    _LATLON: # LatLon
    lambda data: { 
        'Latitude': struct.unpack('<i', data[0:4])[0],
        'Longitude': struct.unpack('<i', data[0:4])[0],
    },
    _AGIOTRAFFIC: # AgIOTraffic
    lambda data: { 
        'Seconds': data[0],
    },
    _IMU: # IMU
    lambda data: { 
        'Heading': struct.unpack('<h', data[0:2])[0]/10.0,
        'Roll': struct.unpack('<h', data[2:4])[0]/10.0
    },
    _IMU_DETACH_REQ: # IMU Detach Req
    lambda data: { # Removed ?
    },
    _NMEA_BYTES: # NMEA bytes
    lambda data: { # ???
    },
    _SWITCH_CONTROL: # switchControl
    lambda data: { 
        'Main': data[0],
        'Res1': data[1],
        'Res2': data[2],
        '# sections': data[3],
        'On Group 0': bin(data[4])[2:],
        'Off Group 0': bin(data[5])[2:],
        'On Group 1': bin(data[6])[2:],
        'Off Group 1': bin(data[7])[2:],
    },
    _MACHINE_CONFIG: # machineConfig
    lambda data: { 
        'raiseTime': data[0],
        'lowerTime': data[1],
        'hydEnable': data[2],
        'set0': data[3],
        'User1': data[4],
        'User2': data[5],
        'User3': data[6],
        'User4': data[7]
    },
    _RELAY_CONFIG: # machineConfig
    lambda data: { 
        'Pin 1': data[0],
        'Pin 2': data[1],
        'Pin 3': data[2],
        'Pin 4': data[3],
        'Pin 5': data[4],
        'Pin 6': data[5],
        'Pin 7': data[6],
        'Pin 8': data[7],
        'Pin 9': data[8],
        'Pin 10': data[9],
        'Pin 11': data[10],
        'Pin 12': data[11],
        'Pin 13': data[12],
        'Pin 14': data[13],
        'Pin 15': data[14],
        'Pin 16': data[15],
        'Pin 17': data[16],
        'Pin 18': data[17],
        'Pin 19': data[18],
        'Pin 20': data[19],
        'Pin 21': data[20],
        'Pin 22': data[21],
        'Pin 23': data[22],
        'Pin 24': data[23]
    },
    _MACHINE_DATA: # machineData
    lambda data: { 
        'uturn': data[0],
        'Speed': data[1]/10.0,
        'hydLift': data[2],
        'Tram': data[3],
        'Geo Stop': data[4],
        'SC': struct.unpack('<H', data[6:8])[0],
    },
    _STEER_CONFIG: # steerConfig
    lambda data: { 
        'set0': data[0],
        'invertWas': data[0] & 1,
        'steerInvertRelays': data[0] >> 1 & 1,
        'invertSteer': data[0] >> 2 & 1,
        'conv': 'Single' if data[0] >> 3 & 1 else 'Differential',
        'motorDrive': 'Cytron' if data[0] >> 4 & 1 else 'IBT2',
        'steerEnable': 'Switch' if data[0] >> 5 & 1 else ('Button' if data[0] >> 6 & 1 else 'None'),
        'encoder': data[0] >> 7 & 1,
        'pulseCount': data[1],
        'minSpeed': data[2],
        'sett1': data[3],
        'danfoss': data[3] & 1,
        'pressureSensor': data[3] >> 1 & 1,
        'currentSensor': data[3] >> 2 & 1,
    },
    _STEERSETTINGS: # steerSettings
    lambda data: { 
        'gainP': data[0],
        'highPWM': data[1],
        'lowPWM': data[2],
        'minPWM': data[3],
        'countsPerDeg': data[4],
        'steerOffset': struct.unpack('<h', data[5:7])[0]/100.0,
        'ackermanFix': data[7],
    },
    _FROM_AUTOSTEER: # fromAutoSteer
    lambda data: { 
        'ActualSteerAngle': struct.unpack('<h', data[0:2])[0]/100.0,
        'IMU Heading Hi/Lo': struct.unpack('<h', data[2:4])[0]/10.0,
        'IMU Roll Hi/Lo': struct.unpack('<h', data[4:6])[0]/10.0,
        'Switch': data[6],
        'PWMDisplay': data[7],
    },
    _AUTOSTEER_DATA: # autoSteerData
    lambda data: { 
        'Speed': struct.unpack('<H', data[0:2])[0]/10.0,
        'Status': data[2],
        'SteerAngle': struct.unpack('<h', data[3:5])[0]/100.0,
        'SC': struct.unpack('<H', data[6:8])[0],
    },
}

alive = [0x80,0x81, 0x7f, 0xC7, 1, 0, 0x47]

class AgIO():
    def __init__(self, settings, motor_control, debug=False):
        self.debug=debug
        self.motor_control=motor_control
        self.settings = settings
        self.sc = pi_steer.section_control.SectionControl()
        self.server=socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
        self.server.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
        self.server.bind(('', 8888))
        self.server.settimeout(0.1)

        self.client=socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
        self.client.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
        self.client.settimeout(0)
        self.client.setblocking(0)

        self.reader_thread = threading.Thread(target=self.read)
        self.reader_thread.start()

    def __del__(self):
        self.reader_thread.join(0)

    def decode_data(self, data):
        try:
            source = data[2]
            pgn = data[3]
            length = data[4]
            crc = data[5+length]
            crc_sum = 0
            for byte in data[2:length+5]:
                crc_sum += byte
            crc_sum %= 256

            crc_text = 'CRC failed.'
            crc_ok = False
            if crc == crc_sum:
                crc_text = 'CRC Ok.'
                crc_ok = True
            else:
                return (None, None)
            
            payload = pgn_data[pgn](data[5:length+5])
            payload_text = '| Payload:'
            for key in payload.keys():
                payload_text += ' ' + key + ' : ' + str(payload[key])
        except KeyError:
            return (None, None)

        if self.debug and pgn not in [0xfe, 0xef, 0xfd]:
            db.write('From: {} | PGN: {} | {} | {}'.format(source_text[source], pgn_text[pgn], payload_text, crc_text))

        if crc_ok:
            self._pgn_handler(pgn, payload)

    def read(self) -> None:
        while True:
            try:
                while True:
                    (data, address)=self.server.recvfrom(1024)
                    if len(data) < 6:
                        continue
                    if data[0] == 0x80 and data[1] == 0x81:
                        self.decode_data(data)
            except socket.error as err:
                if self.debug:
                    db.write('Read socket error: {}'.format(err))
            except socket.timeout as err:
                if self.debug:
                    db.write('Read socket timeout: {}'.format(err))

    def _pgn_handler(self, pgn, payload) -> None:
        if pgn is not None:
            if pgn == _STEER_CONFIG:
                if self.debug:
                    db.write('steer config')
                self.settings.settings.update(payload)
                self.settings.save_settings()
                return
            if pgn == _STEERSETTINGS:
                if self.debug:
                    db.write('steer settings')
                self.settings.settings.update(payload)
                self.settings.save_settings()
                return
            if pgn == _AUTOSTEER_DATA:
                if self.debug:
                    db.write('autosteer data')
                if payload is not None:
                    self.motor_control.set_control(payload)
                    if self.debug:
                        db.write(payload['SC'])
                    self.sc.update(payload['SC'])
                return
            if pgn == _MACHINE_DATA:
                if self.debug:
                    db.write('machine data')
                if payload is not None:
                    self.sc.update(payload['SC'])
                return

    def alive(self):
        self.client.sendto(bytes(alive), ('255.255.255.255',9999))

    def from_autosteer(self, wheel_angle, heading, roll, switch, pwm_display):
        data = bytearray([0x80, 0x81, 0x7e, 0xfd, 0x08])
        wheel_angle_int = int(wheel_angle * 100)
        data.extend(list(struct.pack('<h', wheel_angle_int)))
        heading_int = int(heading * 10)
        data.extend(list(struct.pack('<h', heading_int)))
        roll_int = int(roll * 10)
        data.extend(list(struct.pack('<h', roll_int)))

        data.append(switch)
        data.append(pwm_display)

        crc = 0
        for byte in data[2:]:
            crc += byte
        crc %= 256
        data.append(crc)

        try:
            self.client.sendto(bytes(data), ('255.255.255.255',9999))
        except Exception as err:
            if self.debug:
                db.write('Send error: {}'.format(err))

File: pi-steer-rust/src/communication/agio.rs
use std::net::UdpSocket;
use std::thread;
use std::time::Duration;
use byteorder::{ByteOrder, LittleEndian};
use std::sync::{Arc, Mutex, RwLock};

use crate::config::settings::Settings;

const HELLO: u8 = 0xc7;
const LATLON: u8 = 0xd0;
const AGIOTRAFFIC: u8 = 0xd2;
const IMU: u8 = 0xd3;
const IMU_DETACH_REQ: u8 = 0xd4;
const NMEA_BYTES: u8 = 0xd5;
const SWITCH_CONTROL: u8 = 0xea;
const MACHINE_CONFIG: u8 = 0xee;
const RELAY_CONFIG: u8 = 0xec;
const MACHINE_DATA: u8 = 0xef;
const STEER_CONFIG: u8 = 0xfb;
const STEERSETTINGS: u8 = 0xfc;
const FROM_AUTOSTEER: u8 = 0xfd;
const AUTOSTEER_DATA: u8 = 0xfe;

pub struct Writer {
    client: UdpSocket,
    is_imu: bool,
    debug: bool,
}

impl Writer {
    pub fn new(is_imu: bool, debug: bool) -> Writer {
        let client = UdpSocket::bind("0.0.0.0:0").unwrap();
        client.set_broadcast(true).unwrap();
        println!("Start Writer, IMU {}", is_imu);
        Writer{
            client,
            is_imu,
            debug,
        }
    }

    pub fn from_autosteer(&self, wheel_angle: f32, heading: f32, roll: f32, switch: u8, pwm_value: f64) {
        let mut data = vec![0x80, 0x81, 0x7e, 0xfd, 0x08];
        
        let pwm_display = (pwm_value * 2.55) as u8;
        let wheel_angle_int = (wheel_angle * 100.0) as i16;
        let mut heading_int = ((heading * 10.0) as i16) as u16;
        let mut roll_int = ((roll * 10.0) as i16) as u16;
        if ! self.is_imu {
            heading_int = 9999;
            roll_int = 8888;
        }

        let mut buf = [0; 2];
        LittleEndian::write_i16(&mut buf, wheel_angle_int);
        data.extend_from_slice(&buf);
        LittleEndian::write_u16(&mut buf, heading_int);
        data.extend_from_slice(&buf);
        LittleEndian::write_u16(&mut buf, roll_int);
        data.extend_from_slice(&buf);

        data.push(switch);
        data.push(pwm_display);

        let crc: u8 = data.iter().skip(2).fold(0, |acc, &x| acc.wrapping_add(x));
        data.push(crc);

        match self.client.send_to(&data, "255.255.255.255:9999") {
            Ok(_) => (),
            Err(e) => if self.debug {
                println!("Send error: {:?}", e);
            },
        }
    }

    pub fn gps(&self, _time: &str, lat: f64, ns: &str, lon: f64, ew: &str, fix: u8, sat: u16, hdop: f32, alt: f32, _geoid: &str, age: f32, heading: f32, speed: f32) {
        let mut data: Vec<u8> = vec![0x80, 0x81, 0x7c, 0xd6, 51 as u8];
        let mut buf8 = [0; 8];
        let mut buf4 = [0; 4];
        let mut buf2 = [0; 2];
        let mut longitude = lon;
        if ew == "W" { longitude = -longitude; }
        let mut latitude = lat;
        if ns == "S" { latitude = -latitude; }
        LittleEndian::write_f64(&mut buf8, longitude);
        data.extend_from_slice(&mut buf8);
        LittleEndian::write_f64(&mut buf8, latitude);
        data.extend_from_slice(&mut buf8);
        LittleEndian::write_f32(&mut buf4, f32::MAX); // dual antenna heading
        data.extend_from_slice(&mut buf4);
        LittleEndian::write_f32(&mut buf4, heading); // single antenna heading
        data.extend_from_slice(&mut buf4);
        LittleEndian::write_f32(&mut buf4, speed); // Speed
        data.extend_from_slice(&mut buf4);
        LittleEndian::write_f32(&mut buf4, f32::MAX); // Roll
        data.extend_from_slice(&mut buf4);
        LittleEndian::write_f32(&mut buf4, alt); // Altitude
        data.extend_from_slice(&mut buf4);
        LittleEndian::write_u16(&mut buf2, sat); // Satellites
        data.extend_from_slice(&mut buf2);
        data.push(fix); //Fix
        LittleEndian::write_u16(&mut buf2, (hdop * 100.0) as u16); // HDOP
        data.extend_from_slice(&mut buf2);
        LittleEndian::write_u16(&mut buf2, (age * 100.0) as u16); // Age
        data.extend_from_slice(&mut buf2);
        LittleEndian::write_u16(&mut buf2, u16::MAX); // IMU heading
        data.extend_from_slice(&mut buf2);
        LittleEndian::write_i16(&mut buf2, i16::MAX); // IMU roll
        data.extend_from_slice(&mut buf2);
        LittleEndian::write_i16(&mut buf2, i16::MAX); // IMU pitch
        data.extend_from_slice(&mut buf2);
        LittleEndian::write_i16(&mut buf2, i16::MAX); // IMU yaw
        data.extend_from_slice(&mut buf2);
        let crc: u8 = data.iter().skip(2).fold(0, |acc, &x| acc.wrapping_add(x));
        data.push(crc);

        match self.client.send_to(&data, "255.255.255.255:9999") {
            Ok(_) => (),
            Err(e) => if self.debug {
                println!("Send error: {:?}", e);
            },
        }
    }

    // fn send_heartbeat(client: &UdpSocket) -> std::io::Result<()> {
    //     let heartbeat_message = [0x80, 0x81, 0x7f, HELLO as u8, 1, 0, 0x47];
    //     let broadcast_addr = SocketAddrV4::new(Ipv4Addr::BROADCAST, 9999);
    //     client.send_to(&heartbeat_message, broadcast_addr)?;
    //     println!("Heartbeat message sent.");
    //     Ok(())
    // }
    
}

pub struct PgnData {
    pub sections: RwLock<u16>,
    pub steer_angle: RwLock<f32>,
    pub status: RwLock<bool>,
    pub speed: RwLock<f32>,
}

impl PgnData {
    pub fn get_sections(&self) -> u16 {
        *self.sections.read().unwrap()
    }
}
pub struct Pgn {
    settings: Arc<Mutex<Settings>>,
    pgn_data: Arc<PgnData>,
    debug: bool,
}
pub struct Reader {
}

impl Reader {
    pub fn new(settings: Arc<Mutex<Settings>>, debug: bool) -> Arc<PgnData> {
        let pgn_data = Arc::new(PgnData {
            sections: RwLock::new(0 as u16),
            steer_angle: RwLock::new(0.0 as f32),
            status: RwLock::new(false),
            speed: RwLock::new(0.0 as f32),
        });

        let pgn_clone = Arc::clone(&pgn_data.clone());

        // Reader thread to listen for incoming messages
        thread::spawn(move || Self::reader_thread(settings, pgn_clone, debug));
        pgn_data
    }

    pub fn reader_thread( settings: Arc<Mutex<Settings>>, pgn_data: Arc<PgnData>, debug: bool ) {
        let mut pgn = Pgn { 
            settings, 
            pgn_data,
            debug 
        };

        let server = UdpSocket::bind("0.0.0.0:8888").unwrap();
        server.set_broadcast(true).unwrap();
        server.set_nonblocking(true).unwrap();
        let mut buf = [0u8; 1024];
        loop {
            match server.recv_from(&mut buf) {
                Ok((size, _addr)) => {
                    if size >= 6 {
                        pgn.decode_data(&buf[..size]);
                    }
                },
                Err(e) if e.kind() != std::io::ErrorKind::WouldBlock => {
                    eprintln!("Read socket error: {:?}", e);
                    break;
                },
                _ => {thread::sleep(Duration::from_millis(5));}
            }
        }
    }
}

impl Pgn {

    pub fn new(settings: Arc<Mutex<Settings>>, pgn_data: Arc<PgnData>, debug: bool) -> Pgn {
        Pgn { 
            settings: settings, 
            pgn_data,
            debug 
        }
    }

    pub fn get_speed(self) -> f32 {
        let speed = self.pgn_data.speed.read().unwrap();
        *speed
    }

    pub fn get_status(self) -> bool {
        let status = self.pgn_data.status.read().unwrap();
        *status
    }

    pub fn get_steer_angle(self) -> f32 {
        let steer_angle = self.pgn_data.steer_angle.read().unwrap();
        *steer_angle
    }

    fn decode_data(&mut self, data: &[u8]) -> Option<()> {
        if data.len() < 6 {
            return None; // Early return if data is too short for a valid message.
        }

        let source = data[2];
        let pgn = data[3];
        let length = data[4] as usize;

        if data.len() < length + 6 { return None; } // Check if the length is valid.

        let crc = data[5 + length];
        let crc_sum: u8 = data[2..length + 5].iter().fold(0u8, |acc, &x| acc.wrapping_add(x));

        let crc_ok = crc == crc_sum;

        // Use debug_write for debug messages, assuming it's a stand-in for the db.write function.
        if self.debug && pgn != 0xfe && pgn != 0xef && pgn != 0xfd {
            let crc_text = if crc_ok { "CRC Ok." } else { "CRC failed." };
            println!("From: {} | PGN: {} | CRC: {}", source, pgn, crc_text);
        }

        if !crc_ok {
            return None;
        }

        self.pgn_handler(pgn, &data[5..5 + length]);

        Some(())
    }

    fn pgn_handler(&mut self, pgn: u8, data: &[u8]) {
        match pgn {
            HELLO => {},
            LATLON => {},
            AGIOTRAFFIC => {}
            IMU => {},
            IMU_DETACH_REQ => {},
            NMEA_BYTES => {},
            SWITCH_CONTROL => {},
            MACHINE_CONFIG => {},
            RELAY_CONFIG => {},
            MACHINE_DATA => {
                // let uturn = data[0];
                // let speed = data[1] as f32 / 10.0;
                // let hyd_lift = data[2];
                // let tram = data[3];
                // let geo_stop = data[4];
                let sc = LittleEndian::read_u16(&data[6..8]);
        
                if self.debug {
                    println!("machine data");
                }
                let mut sections = self.pgn_data.sections.write().unwrap();
                *sections = sc;
            },
            STEER_CONFIG => {
                let mut settings = self.settings.lock().unwrap();
                // let set0 = data[0];
                settings.invert_was = data[0] & 1 == 1;
                settings.steer_invert_relays = data[0] >> 1 & 1 == 1;
                settings.invert_steer = data[0] >> 2 & 1 == 1;
                settings.conv = if data[0] >> 3 & 1 == 1 { "Single".to_string() } else { "Differential".to_string() };
                settings.motor_drive = if data[0] >> 4 & 1 == 1 { "Cytron".to_string() } else { "IBT2".to_string() };
                settings.steer_enable = if data[0] >> 5 & 1 == 1 { "Switch".to_string() } else { if data[0] >> 6 & 1 == 1 { "Button".to_string() } else { "None".to_string() } };
                settings.encoder = data[0] >> 7 & 1 == 1;
                // let pulse_count = data[1];
                // let min_speed = data[2];
                // let sett1 = data[3];
                settings.danfoss = data[3] & 1 == 1;
                settings.pressure_sensor = data[3] >> 1 & 1 == 1;
                settings.current_sensor = data[3] >> 2 & 1 == 1;
        
                if self.debug {
                    println!("steer config");
                }
                settings.save_settings();
            },
            STEERSETTINGS => {
                let mut settings = self.settings.lock().unwrap();
                settings.gain_p = data[0];
                settings.high_pwm = data[1];
                settings.low_pwm = data[2];
                settings.min_pwm = data[3];
                settings.counts_per_deg = data[4];
                settings.steer_offset = LittleEndian::read_i16(&data[5..7]) as f32 / 100.0;
                settings.ackerman_fix = data[7];

                if self.debug {
                    println!("steer settings");
                }
                settings.save_settings();
            },
            FROM_AUTOSTEER => {},
            AUTOSTEER_DATA => {
                let speed = LittleEndian::read_u16(&data[0..2]) as f32 / 10.0;
                let status = data[2];
                let steer_angle = LittleEndian::read_i16(&data[3..5]) as f32 / 100.0;
                let sc = LittleEndian::read_u16(&data[6..8]);

                if self.debug {
                    println!("autosteer data");
                    println!("SC: {:#018b}, steer angle: {}", sc, steer_angle);
                }
                let mut _steer_angle = self.pgn_data.steer_angle.write().unwrap();
                *_steer_angle = steer_angle;
                let mut _status = self.pgn_data.status.write().unwrap();
                *_status = status != 0;
                let mut _speed = self.pgn_data.speed.write().unwrap();
                *_speed = speed;
                let mut sections = self.pgn_data.sections.write().unwrap();
                *sections = sc;
            },
            _ => (),
        }
    }

}

#[test]
fn settest() {

}
